src/memo/memo.ml

```diff
@@ -685,254 +638,413 @@ let create_hidden (type output) name ?doc ~input typ impl =
+    ('a, 'b, 'a -> 'b Fiber.t) Dep_node.t -> 'b Cached_value.t Fiber.t
+
+  (** [exec] and variants thereof *)
   val exec : ('a, 'b, 'a -> 'b Fiber.t) t -> 'a -> 'b Fiber.t
+
+  val exec_dep_node : ('a, 'b, 'a -> 'b Fiber.t) Dep_node.t -> 'b Fiber.t
 end = struct
-  let compute inp ivar (dep_node : _ Dep_node.t) running_state =
-    (* define the function to update / double check intermediate result *)
-    (* set context of computation then run it *)
+  module Start_considering_result = struct
+    type 'a t =
+      | Done of 'a Cached_value.t
+      | Need_work of
+          { ivar : 'a Cached_value.t Fiber.Ivar.t
+          ; sample_attempt : Dag.node
+                (** [work] must be called to force [ivar] to be filled *)
+          ; work : unit Fiber.t option
+          }
+
+    let sample_attempt_dag_node t : Sample_attempt_dag_node.t =
+      match t with
+      | Done _ -> Finished
+      | Need_work { sample_attempt; _ } -> Running sample_attempt
+  end
+
+  let deps_changed =
+    let rec go deps =
+      match deps with
+      | [] -> Fiber.return Changed_or_not.Unchanged
+      | Last_dep.T (dep, cv) :: deps -> (
+        let* res = exec_dep_node_internal dep in
+        match Value_id.equal res.id cv.id with
+        | true -> go deps
+        | false -> Fiber.return Changed_or_not.Changed )
+    in
+    go
+
+  module Prev_cycle_cache_lookup_result = struct
+    type ('a, 'b) t =
+      | Valid of 'a
+      | Invalid of { old_value : 'b option }
+  end
+
+  let do_validate (dep_node : _ Dep_node.t) inp ivar running_state =
     let* res =
       Call_stack.push_async_frame
         (T { without_state = dep_node.without_state; running_state })
         (fun () ->
-          match dep_node.without_state.spec.f with
-          | Function.Async f ->
-            (* A consequence of using [Fiber.collect_errors] is that memoized
-               functions don't report errors promptly - errors are reported once
-               all child fibers terminate. To fix this, we should use
-               [Fiber.with_error_handler], but we dont have access to dune's
-               error reoprting mechanism in memo *)
-            Fiber.collect_errors (fun () -> f inp))
-    in
-    (* update the output cache with the correct value *)
-    let deps = List.rev running_state.deps_so_far.deps_reversed in
-    let res =
-      Result.map_error res ~f:(fun exns ->
-          (* this step deduplicates the errors *)
-          Value.Async (Exn_set.of_list exns))
+          let* from_cache =
+            match dep_node.last_cached_value with
+            | None ->
+              Fiber.return
+                (Prev_cycle_cache_lookup_result.Invalid { old_value = None })
+            | Some cv -> (
+              let+ res = deps_changed cv.deps in
+              match res with
+              | Unchanged ->
+                dep_node.last_cached_value <-
+                  Some { cv with last_validated_at = Run.current () };
```

Comment:
Bug? `last_validated_at` probably needs to be mutated, not the whole `last_cached_value`.

Reply:
I changed this to do the mutation, but it was not a correctness issue because nothing inspects the `last_validated_at` field of `Cached_value`s stored in `Last_dep` (and that's the only way the difference could be observable, I think).

I changed `Last_dep` to only inlclude `Value_id` instead of including the whole `Cached_value` so it's more clear now.